subsys/bluetooth/host/scan.c

```diff
@@ -639,22 +657,50 @@ void bt_hci_le_per_adv_report(struct net_buf *buf)
 	info.addr = &per_adv_sync->addr;
 	info.sid = per_adv_sync->sid;
 
-	if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL) {
-		/* Handling of incomplete reports is currently not
-		 * handled in the host. The remaining advertising
-		 * reports may therefore contain partial data.
-		 */
-		BT_WARN("Incomplete per adv report");
-	}
-
-	SYS_SLIST_FOR_EACH_CONTAINER(&pa_sync_cbs, listener, node) {
-		if (listener->recv) {
-			net_buf_simple_save(&buf->b, &state);
-
-			buf->len = evt->length;
-			listener->recv(per_adv_sync, &info, &buf->b);
-
-			net_buf_simple_restore(&buf->b, &state);
+	if (!per_adv_sync->report_truncated) {
+		if (per_adv_sync->reassembly == NULL) {
+			if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE) {
+				/* No reassembly buffer and a complete report.
+				 * Forward it without an extra copy.
+				 */
+				bt_hci_le_per_adv_report_recv(per_adv_sync, buf, &info);
+			} else {
+				BT_WARN("No reassembly buffer. Dropping the report");
+				per_adv_sync->report_truncated = false;
```

Comment:
So the design here is that if we are already assembling another report, any future reports will be discarded? Shouldn't it be the other way, i.e. if we have a report R1 that is not fully reassembled yet, and we receive a new report R2 from the same periodic advertiser, shouldn't we then discard R1, as we probably shouldn't expect more segments from R1?

Reply:
I'm not sure I understand what you mean.
The host will receive a sequence of ```BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL``` reports followed by one with the ```BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE``` flag set.
All of those belong together and if any of them cannot be put in the buffer then we need to wait for that last ```BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE``` report and discard them all.

So there should not be two separate reports at all.

Reply:
If the case is that we will always get 0 or more `BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL` followed by a `BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE`, then when will this where `BT_WARN("No reassembly buffer. Dropping the report");` ever happen?

Reply:
It will happen if the allocation on line 807 fails for some reason or if ```CONFIG_BT_PER_ADV_SYNC_BUF_SIZE == 0```.

It is possible for all receivers to share one bigger pool if that is the preferred behavior.

Reply:
Right, so if the suggestion from https://github.com/zephyrproject-rtos/zephyr/pull/40596#discussion_r760178117 is implemented, this error handling could be removed.